File: agents/src/middleware/jwt.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type { Request, Response, NextFunction, RequestHandler } from "express";
import { createRemoteJWKSet, jwtVerify, errors as joseErrors } from "jose";

export interface JWTAuthOptions {
  /** Thunder/IDP JWKS endpoint. */
  jwksUrl: string;
  /** Acceptable iss claim value. */
  issuer: string | string[];
  /** Acceptable aud claim value. */
  audience: string;
  /**
   * RFC 9728 protected-resource metadata URL. Included in the
   * WWW-Authenticate Bearer challenge when set.
   */
  resourceMetadataUrl?: string;
}

const REALM = "asdlc";

function buildBearerChallenge(resourceMetadataUrl: string | undefined, errorCode: string): string {
  const parts: string[] = [`realm="${REALM}"`];
  if (errorCode) parts.push(`error="${errorCode}"`);
  if (resourceMetadataUrl) parts.push(`resource_metadata="${resourceMetadataUrl}"`);
  return `Bearer ${parts.join(", ")}`;
}

/**
 * Express middleware that validates a JWT against a remote JWKS.
 *
 * Uses jose.createRemoteJWKSet, which has a built-in cache and coalesces
 * concurrent fetches. On a kid miss it triggers one refresh before failing —
 * matches the behaviour of the Go middleware.
 *
 * Verified claims are attached to res.locals.tokenClaims; the raw token to
 * res.locals.token. Both fields are typed via the express namespace
 * augmentation in src/types/express.d.ts (or the consuming service's
 * equivalent) — but kept loose here to avoid a hard module dependency.
 */
export function jwtAuthMiddleware(opts: JWTAuthOptions): RequestHandler {
  const jwks = createRemoteJWKSet(new URL(opts.jwksUrl));

  return async function jwtAuth(req: Request, res: Response, next: NextFunction): Promise<void> {
    const auth = req.header("Authorization");
    if (!auth) {
      res.setHeader("WWW-Authenticate", buildBearerChallenge(opts.resourceMetadataUrl, ""));
      res.status(401).json({ error: "missing Authorization header" });
      return;
    }
    const m = /^Bearer\s+(.+)$/i.exec(auth);
    if (!m) {
      res.setHeader("WWW-Authenticate", buildBearerChallenge(opts.resourceMetadataUrl, "invalid_token"));
      res.status(401).json({ error: "malformed Authorization header" });
      return;
    }
    const token = m[1].trim();

    try {
      const { payload, protectedHeader } = await jwtVerify(token, jwks, {
        issuer: opts.issuer,
        audience: opts.audience,
        algorithms: ["RS256"],
      });
      // jose.jwtVerify already validates iss, aud, exp, nbf when the options
      // are passed. We only need to surface the verified claims downstream.
      (res.locals as Record<string, unknown>).tokenClaims = payload;
      (res.locals as Record<string, unknown>).tokenHeader = protectedHeader;
      (res.locals as Record<string, unknown>).token = token;
      next();
    } catch (err) {
      const code = err instanceof joseErrors.JOSEError ? err.code : "invalid_token";
      const msg = err instanceof Error ? err.message : "invalid jwt";
      // eslint-disable-next-line no-console
      console.warn("jwt validation failed", { code, msg });
      res.setHeader("WWW-Authenticate", buildBearerChallenge(opts.resourceMetadataUrl, "invalid_token"));
      res.status(401).json({ error: "invalid jwt" });
    }
  };
}

File: services/aep-api/internal/platform/auth/jwt.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

// The BFF inbound JWT identity layer. It wraps the jwtassertion subpackage
// (JWKS-backed RS256 verification) and exposes a thin Claims projection for the
// rest of the service to consume.
package auth

import (
	"context"
	"log/slog"
	"net/http"
	"strings"

	"github.com/wso2/aep/aep-api/internal/authz"
	"github.com/wso2/aep/aep-api/internal/platform/auth/jwtassertion"
)

// Claims is the BFF-internal projection of the verified JWT.
type Claims struct {
	Subject  string
	ClientID string
	// Organisation claims sourced from Thunder. ResolveOuHandle is the
	// only place that picks one — keep the precedence in lockstep with
	// the console.
	OuHandle string
	OuName   string
	OuId     string
	// Scope is the raw space-delimited OAuth scope claim. AE permission
	// keys (e.g. "ae:build") arrive as entries in this string alongside
	// unrelated scopes (openid, profile, system, ...); see Permissions.
	Scope string
}

// ResolveOuHandle returns the canonical OC org handle from a verified
// JWT, preferring `ouHandle` over `ouName` over `ouId`. Returns "" when
// the token has none of those claims (which the caller must surface as
// a fail-loud error rather than silently substitute an org).
//
// The console mirrors this precedence verbatim
// (console/src/utils/orgClaims.ts). Any change here MUST land on both
// sides simultaneously.
func ResolveOuHandle(c *Claims) string {
	if c == nil {
		return ""
	}
	if c.OuHandle != "" {
		return c.OuHandle
	}
	if c.OuName != "" {
		return c.OuName
	}
	return c.OuId
}

func (c *Claims) Permissions() []authz.Permission {
	if c == nil {
		return nil
	}
	return filterPermissions(c.Scope)
}

func filterPermissions(scope string) []authz.Permission {
	known := make(map[authz.Permission]struct{}, len(authz.AllPermissions))
	for _, p := range authz.AllPermissions {
		known[p] = struct{}{}
	}

	var perms []authz.Permission
	for _, token := range strings.Fields(scope) {
		p := authz.Permission(token)
		if _, ok := known[p]; ok {
			perms = append(perms, p)
		}
	}
	return perms
}

type claimsContextKey struct{}

// WithClaims returns a copy of ctx that carries the given Claims value.
func WithClaims(ctx context.Context, claims *Claims) context.Context {
	return context.WithValue(ctx, claimsContextKey{}, claims)
}

// ClaimsFromContext retrieves the Claims stored by WithClaims.
func ClaimsFromContext(ctx context.Context) *Claims {
	c, _ := ctx.Value(claimsContextKey{}).(*Claims)
	return c
}

// JWTConfig configures the inbound JWT verifier. It is a thin shim over
// jwtassertion.Config with the same fields.
type JWTConfig struct {
	JWKS                *jwtassertion.JWKSCache
	AllowedIssuers      []string
	AllowedAudiences    []string
	ResourceMetadataURL string
}

// JWTMiddleware returns an HTTP middleware that verifies the Authorization
// header against cfg, projects the verified token into a Claims record, and
// forwards the request with the projection in context.
//
// The underlying jwtassertion middleware emits RFC 9728 WWW-Authenticate
// challenges on failure. The thin projection only exists so the rest of
// the BFF doesn't have to care about the full TokenClaims shape.
func JWTMiddleware(cfg JWTConfig) func(http.Handler) http.Handler {
	verifier := jwtassertion.Authenticator(jwtassertion.Config{
		JWKS:                cfg.JWKS,
		AllowedIssuers:      cfg.AllowedIssuers,
		AllowedAudiences:    cfg.AllowedAudiences,
		ResourceMetadataURL: cfg.ResourceMetadataURL,
	})
	return func(next http.Handler) http.Handler {
		return verifier(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			tc := jwtassertion.GetTokenClaims(r.Context())
			if tc != nil {
				// tc.Sub, not tc.Subject: TokenClaims declares its own
				// `Sub string json:"sub"` at depth 0, which shadows the
				// embedded jwt.RegisteredClaims.Subject (same tag, deeper) —
				// encoding/json decodes ONLY the shallower field, so the
				// promoted Subject is always empty on this edge (found via
				// e2e: turn commits fell back to the credential identity).
				claims := &Claims{
					Subject:  tc.Sub,
					ClientID: tc.ClientID,
					OuHandle: tc.OuHandle,
					OuName:   tc.OuName,
					OuId:     tc.OuId,
					Scope:    tc.Scope,
				}
				ctx := WithClaims(r.Context(), claims)
				r = r.WithContext(ctx)

				// Debug-only visibility into which AE permission keys this
				// request's token actually carries — the same derivation
				// permission_gate.go checks against, not a re-parse of the
				// raw scope, so this can never drift from what the gate
				// decides. Deliberately omits the raw Scope/token: only the
				// filtered, known AE keys are logged.
				slog.DebugContext(ctx, "console request authenticated",
					"subject", claims.Subject,
					"org", ResolveOuHandle(claims),
					"permissions", claims.Permissions(),
					"method", r.Method, "path", r.URL.Path)
			}
			next.ServeHTTP(w, r)
		}))
	}
}
